src/audio/template_comp/template.c

```diff
@@ -0,0 +1,213 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2025 Intel Corporation.
+
+#include <sof/audio/module_adapter/module/generic.h>
+#include <sof/audio/sink_source_utils.h>
+#include <sof/audio/sink_api.h>
+#include <sof/audio/source_api.h>
+#include <rtos/init.h>
+#include "template.h"
+
+/* UUID identifies the components. Use e.g. command uuidgen from package
+ * uuid-runtime, add it to uuid-registry.txt in SOF top level.
+ */
+SOF_DEFINE_REG_UUID(template_comp);
+
+/* Creates logging data for the component */
+LOG_MODULE_REGISTER(template_comp, CONFIG_SOF_LOG_LEVEL);
+
+/* Creates the compont trace. Traces show in trace console the component
+ * info, warning, and error messages.
+ */
+DECLARE_TR_CTX(template_comp_tr, SOF_UUID(template_comp_uuid), LOG_LEVEL_INFO);
+
+/**
+ * template_comp_init() - Initialize the template component.
+ * @mod: Pointer to module data.
+ *
+ * This function is called when the instance is created. The
+ * macro __cold informs that the code that is non-critical
+ * is loaded to slower but large DRAM.
+ *
+ * Return: Zero if success, otherwise error code.
+ */
+__cold static int template_comp_init(struct processing_module *mod)
+{
+	struct module_data *md = &mod->priv;
+	struct comp_dev *dev = mod->dev;
+	struct template_comp_comp_data *cd;
+
+	comp_info(dev, "template_comp_init()");
+
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*cd));
+	if (!cd)
+		return -ENOMEM;
+
+	md->private = cd;
+	return 0;
+}
+
+/**
+ * template_comp_process() - The audio data processing function.
+ * @mod: Pointer to module data.
+ * @sources: Pointer to audio samples data sources array.
+ * @num_of_sources: Number of sources in the array.
+ * @sinks: Pointer to audio samples data sinks array.
+ * @num_of_sinks: Number of sinks in the array.
+ *
+ * This is the processing function that is called for scheduled
+ * pipelines. The processing is controlled by the enable switch.
+ *
+ * Return: Zero if success, otherwise error code.
+ */
+static int template_comp_process(struct processing_module *mod,
+				 struct sof_source **sources,
+				 int num_of_sources,
+				 struct sof_sink **sinks,
+				 int num_of_sinks)
+{
+	struct template_comp_comp_data *cd =  module_get_private_data(mod);
+	struct comp_dev *dev = mod->dev;
+	struct sof_source *source = sources[0];
+	struct sof_sink *sink = sinks[0];
+	int frames = source_get_data_frames_available(source);
+	int sink_frames = sink_get_free_frames(sink);
+	int ret = 0;
+
+	comp_dbg(dev, "template_comp_process()");
+
+	frames = MIN(frames, sink_frames);
+	if (cd->enable) {
+		/* Process the data with the channels swap example function. */
+		ret = cd->template_comp_func(mod, source, sink, frames);
```

Comment:
could do `return cd->template_comp_func(...);` and remove the `else` below and the `ret` variable